DA-933: add huggingface gsi tutorial

VirajAgarwal-couchbase opened the pull request:
- Updated GSI & FTS tutorial for HuggingFace

VirajAgarwal1 committed:
update: make dirs for fts and gsi

VirajAgarwal1 committed:
update: fts tutorial's dependencies

VirajAgarwal1 committed:
updated: intro and frontmatter to explicitly mention fts or gsi

VirajAgarwal1 committed:
updated: frontmatter for gsi

VirajAgarwal1 committed:
added: notebook for gsi tutorial for huggingface

VirajAgarwal1 committed:
update: execution results in the fts and gsi tutorials

prajwal-pai77 commented on huggingface/fts/formatter.md:
```diff
@@ -0,0 +1,20 @@
+---
+# frontmatter
+path: "/tutorial-huggingface-couchbase-vector-search-using-fts"
```
For standardized naming, we can keep it -with-fts

prajwal-pai77 commented on huggingface/fts/formatter.md:
```diff
@@ -0,0 +1,20 @@
+---
+# frontmatter
```
Shouldn't the file be called frontmatter.md

VirajAgarwal-couchbase replied:
It is called `frontmatter.md` That comment is present in all the other tutorials, why would we need to add `.md` in front of it?

prajwal-pai77 replied:
I mean to say the file name, I can see its called formatter not frontmatter

VirajAgarwal-couchbase replied:
Ohh, I see
Thank you for this! 👍 I didn't notice this problem in hugging face tutorials while I was editing the files.

prajwal-pai77 commented on huggingface/fts/hugging_face.ipynb:
```diff
@@ -7,7 +7,7 @@
     "\n",
-    "In this guide, we will walk you through building a powerful semantic search engine using Couchbase as the backend database, [Hugging Face](https://huggingface.co/) as the AI-powered embedding Model. Semantic search goes beyond simple keyword matching by understanding the context and meaning behind the words in a query, making it an essential tool for applications that require intelligent inf ...
+    "In this guide, we will walk you through building a powerful semantic search engine using Couchbase as the backend database, [Hugging Face](https://huggingface.co/) as the AI-powered embedding Model. Semantic search goes beyond simple keyword matching by understanding the context and meaning behind the words in a query, making it an essential tool for applications that require intelligent inf ...
    ]
```
Pls point it to the developer tutorial

prajwal-pai77 commented on huggingface/fts/hugging_face.ipynb:
```diff
@@ -101,18 +104,21 @@
-    "couchbase_scope = input(\"Couchbase Scope:\")\n",
-    "couchbase_collection = input(\"Couchbase Collection:\")"
+    "# Load environment variables\n",
+    "load_dotenv(\"./.env\")\n",
```
Pls have a .env.sample file like in other tutorial, so users can get it from there

prajwal-pai77 commented on huggingface/gsi/formatter.md:
```diff
@@ -1,12 +1,12 @@
-path: "/tutorial-huggingface-couchbase-vector-search"
-title: Using Hugging Face Embeddings with Couchbase Vector Search
-short_title: Hugging Face with Couchbase Vector Search
+path: "/tutorial-huggingface-couchbase-vector-search-using-global-secondary-indexes"
```
For standardizing, keep it -with-global-secondary-index

prajwal-pai77 commented on huggingface/gsi/hugging_face.ipynb:
```diff
@@ -0,0 +1,518 @@
+    "\n",
+    "This guide is designed to be comprehensive yet accessible, with clear step-by-step instructions that will equip you with the knowledge to create a fully functional semantic search system. Whether you're building a recommendation engine, content discovery platform, or any application requiring intelligent document retrieval, this tutorial provides the foundation you need.\n",
+    "\n",
+    "**Note**: If you want to perform semantic search using the FTS (Full-Text Search) index instead, please take a look at [this alternative approach](../fts/hugging_face.ipynb)."
```
Pls point to developer portal

prajwal-pai77 commented on huggingface/gsi/hugging_face.ipynb:
```diff
@@ -0,0 +1,518 @@
+   "outputs": [],
+   "source": [
+    "# Load environment variables\n",
+    "load_dotenv(\"./.env\")\n",
```
Pls have a .env.sample file like in other tutorial, so users can get it from there

VirajAgarwal-couchbase replied:
Added the `.env.sample` file in both fts and gsi

prajwal-pai77 commented on huggingface/gsi/hugging_face.ipynb:
```diff
@@ -0,0 +1,518 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Create BHIVE index if needed\n",
```
Any reason we have this comment?

VirajAgarwal-couchbase replied:
The `if needed` got added by mistake. Thank you for pointing this out.

prajwal-pai77 commented on huggingface/fts/hugging_face.ipynb:
```diff
@@ -346,6 +344,14 @@
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
```
Pls remove empty cell

prajwal-pai77 commented on huggingface/gsi/hugging_face.ipynb:
```diff
@@ -0,0 +1,518 @@
+   "id": "f11a0d98-bcf5-4fe4-b602-6e8a23edf95e",
+   "metadata": {},
+   "source": [
+    "# Performing Semantic Search with GSI Vector Index\n",
```
Would it be a good idea to show the speed up before and after creating the bhive index?

prajwal-pai77 reviewed (changes requested):
Nice work! Some changes & should be good to go

VirajAgarwal1 committed:
added: env sample files

VirajAgarwal1 committed:
update: spelling mistake in frontmatter and path

VirajAgarwal1 committed:
updated: intro links to point to dev portal instead

VirajAgarwal1 committed:
update: gsi frontmatter path

prajwal-pai77 reviewed (changes requested):
As discussed, couple of changes:
- Change score to distance everywhere
- Move couchbase Cache to RAG section
- Add RAG section similar to other tutorials

VirajAgarwal1 committed:
update: based on comments by Prajwal

- restructure hugging_face notebook with improved section headings
- detailed explanations for GSI vector search and embedding processes
- perfomance comparision
- changed score to distance

nithishr commented on huggingface/gsi/hugging_face.ipynb:
There are a couple of issues:
* Vector Search with FTS is available from CB Server 7.6 & not 7.6.4
* RAG search function is not quite optimal
  - The results should include the documents & there should not be a separate call to fetch the doc via KV, which adds overhead.
* Also, the comparison between Baseline Vector Search (without index), Caching & Vector Search with GSI index is not quite right. 
 - Caching can be applied on top of any Vector Search to improve the performance. Please correct me if this is not the case with GSI + Caching.

nithishr replied:
Also mention about the alternative Index type, Composite in case people want to try it out

VirajAgarwal-couchbase replied:
Have made the changes based on your comments. Thank you Nithish 👍

VirajAgarwal-couchbase replied:
Caching does seem to work on FTS and GSI both, when looking at response times. But am not sure implementation-wise.

VirajAgarwal1 committed:
update: according to Nithish's comments

- Added more explanatrion for Composite Index
- Updated performance comparison
- corrected version info in gsi tutorial
